components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java

```diff
@@ -1962,4 +1968,70 @@ public static boolean isSCIM2UserMaxItemsPerPageEnabled() {
         }
         return Boolean.parseBoolean(scim2UserMaxItemsPerPageEnabledProperty);
     }
+
+    /**
+     * Validates the signature of the given data for the specified tenant domain.
+     *
+     * @param data         The data to be verified.
+     * @param signature    The signature to be verified.
+     * @param tenantDomain The tenant domain to which the data belongs.
+     * @return true if the signature is valid, false otherwise.
+     * @throws IdentityKeyStoreResolverException If an error occurs during the signature validation process.
+     */
+    public static boolean validateSignature(String data, byte[] signature, String tenantDomain)
+            throws IdentityKeyStoreResolverException {
+
+        try {
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+            IdentityTenantUtil.initializeRegistry(tenantId);
+        } catch (IdentityException e) {
+            throw new IdentityKeyStoreResolverException(
+                    IdentityKeyStoreResolverConstants.ErrorMessages
+                            .ERROR_CODE_ERROR_RETRIEVING_TENANT_PRIVATE_KEY.getCode(),
+                    "Error while loading the private key", e);
+        }
+        PublicKey publicKey = IdentityKeyStoreResolver.getInstance().getCertificate(tenantDomain, null)
+                .getPublicKey();
+        return SignatureUtil.validateSignature(data, signature, publicKey);
+    }
+
+    /**
+     * Sign the given data for the specified tenant domain.
+     *
+     * @param data         The data to be signed.
+     * @param tenantDomain The tenant domain to which the data belongs.
+     * @return The signature of the data.
+     * @throws IdentityKeyStoreResolverException If an error occurs during the signature generation process.
+     */
+    public static byte[] doSignature(String data, String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        PrivateKey privateKey;
+
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                privateKey = keyStoreManager.getDefaultPrivateKey();
+            } else {
+                String tenantKeyStoreName = IdentityKeyStoreResolverUtil.buildTenantKeyStoreName(tenantDomain);
```

Comment:
We can move the `catch (Exception e)` to here

Reply:
A different change is sent from this commit - https://github.com/wso2/carbon-identity-framework/pull/6176/commits/43523e575a591197e6f3b04210f9be0a83ee4fe0